### PR TITLE
Use random_attention_mask for TF tests

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

- Change TF's `random_attention_mask` to match its PT/Flax equivalence.
- Use `random_attention_mask` defined in `test_modeling_tf_common.py` to generate attention mask in TF tests.

  - so TF code has the same logic as in PT/Flax tests (regarding this attention mask part in tests)
  - avoid large difference between PT/TF outputs. (In particular, `TFGPT2EncoderDecoderModelTest` in [here](https://github.com/huggingface/transformers/issues/16497))
    - In the case of ``TFBERTEncoderDecoderModelTest`` or `TFGPT2EncoderDecoderModelTest`, it is caused by some sequence in a batch which gets all 0s as attention mask (generated by `ids_tensor`) - may happens on both encoder and decoder (especially after combining with the causal mask).

## More context

Currently, most of TF tests still uses
```python
input_mask = ids_tensor([self.batch_size, self.seq_length], vocab_size=2)
```
while in PT/Flax tests, they call
```
input_mask = random_attention_mask([self.batch_size, self.seq_length])
```
(defined in the comment test file).

In particular, `random_attention_mask` has
```
    # make sure that at least one token is attended to for each batch
    attn_mask[:, -1] = 1
```

### CODE REVIEW

File: tests/clip/test_modeling_tf_clip.py [hunk 1]
```diff
@@ -301,6 +301,12 @@ def prepare_config_and_inputs(self):
         input_mask = None
         if self.use_input_mask:
             input_mask = random_attention_mask([self.batch_size, self.seq_length])
+            # make sure the first token has attention mask `1` to ensure that, after combining the causal mask, there
+            # is still at least one token being attended to for each batch.
+            # TODO: Change `random_attention_mask` in PT/TF/Flax common test file, after a discussion with the team.
+            input_mask = tf.concat(
```
Comment (ydshieh) [inline tests/clip/test_modeling_tf_clip.py:307]: This is added to make TF CLIP pass.
(as TF's `random_attention_mask` is changed too in this PR)

File: tests/test_modeling_tf_common.py [hunk 2]
```diff
@@ -1440,7 +1440,7 @@ def ids_tensor(shape, vocab_size, rng=None, name=None, dtype=None):
 def random_attention_mask(shape, rng=None, name=None, dtype=None):
     attn_mask = ids_tensor(shape, vocab_size=2, rng=None, name=None, dtype=dtype)
     # make sure that at least one token is attended to for each batch
-    attn_mask = tf.concat([tf.constant(value=1, shape=(shape[0], 1), dtype=dtype), attn_mask[:, 1:]], axis=1)
+    attn_mask = tf.concat([attn_mask[:, :-1], tf.ones_like(attn_mask[:, -1:], dtype=dtype)], axis=-1)
```
Comment (ydshieh) [inline tests/test_modeling_tf_common.py:1443]: This is changed to match PT/Flax's `random_attention_mask`.

**Review by sgugger — APPROVED** [review]
Thanks for making this more consistent with the rest of the library!

**Review by gante — APPROVED** [review]
Interesting. It moves the column of 1s for the start to the end and now becomes like a left-padded input. It could help with GPT-2, indeed

**Comment by ydshieh** [comment]
> Interesting. It moves the column of 1s for the start to the end and now becomes like a left-padded input. It could help with GPT-2, indeed

(If you are interesting to know a bit the detail, @gante )

Actually, moving 1 to the end will cause problem (when the model use causal mask.). This is why I needed to update the code in `TFCLIPModelTest`.

In general, current library has a bit issue when the final attention mask (after combining the causal mask if any) received by the attention layer has a sequence (in the batch) having all 0s as mask. One thing (but maybe not only) involved is the different values (-1e4, -1e9, -1e30, -inf) used.

Put 1 at the start will avoid this situation (when combining the causal mask).
(But I don't want to change the PT/Flax logic in this PR. This should be addressed in a separate PR after discussion.)

**Review by Rocketknight1 — APPROVED** [review]
Writing as I go to make sure I follow:

- TF tests used to call `ids_tensor` with a vocabulary of 2 to generate a random attention mask
- They now call `random_attention_mask`, which also generates a tensor containing 0 and 1 only, but guarantees that at least one token will have a value of 1.
- This matches the behaviour in the rest of the library, guarantees we will never get a fully-masked input, and slightly increases the expected number of unmasked tokens in each input.

Seems like a great change for both test reliability and consistency with the rest of the library!

**Comment by ydshieh** [comment]
Hi, @Rocketknight1,

Yes, all the points are right -- except

- I am not sure about this statement `and slightly increases the expected number of unmasked tokens in each input.`: I would say not this case, but I might misunderstand the sentence.

- `but guarantees that at least one token will have a value of 1`:
  - Yes, but not guarantee the same thing for the final attention mask used by attention layers to compute the softmax - because the final mask might be the one after combining the causal mask (for decoder models).
  - Some more future PRs to improve these kinds of things.

**Comment by Rocketknight1** [comment]
That makes sense! And my comment about "increases the expected number of unmasked tokens" was just an irrelevant observation - the average number of unmasked tokens is very slightly larger since we guarantee that one of them will have value 1. Ignore me!
